README.md
# Mibe IoT Mirror

IoT Mirror

## Requirements

* device Raspberry Pi Zero W
* installed Raspberry Pi OS (Legacy)
* BlueZ >= 5.47
* Node.js >= 12.22.0 && npm >= 6.5.2
* installed `libdbus-1-dev` and `libglib2.0-dev` dependencies

## How to run

1. `git clone https://github.com/mibe-iot/mirror && cd mirror`
2. `sudo mv resources/mirrorApp.conf /etc/dbus-1/system.d/`
3. `npm i`
4. `node src/index`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

### [2.1.5](https://github.com/mibe-iot/mirror/compare/v2.1.4...v2.1.5) (2022-05-17)


### Bug Fixes

* **pairing:** remove ble paired attribute ([1556bd9](https://github.com/mibe-iot/mirror/commit/1556bd98af5dbef5a4c431424e480ceb2aba7e68))
* **screen:** mqtt subscribe to all ([b328787](https://github.com/mibe-iot/mirror/commit/b328787dd2f0d5ab3103bf84a1cce698b6617050))

### [2.1.4](https://github.com/mibe-iot/mirror/compare/v2.1.3...v2.1.4) (2022-05-16)


### Bug Fixes

* **env:** add default value for `DBUS_SERVICE_NAME` ([12e766c](https://github.com/mibe-iot/mirror/commit/12e766c07aefabe889a4cc53fdb42138905cc0a4))
* **resources:** add `dbus` config for application ([c01818b](https://github.com/mibe-iot/mirror/commit/c01818be68140b4a647bcad1a5f281526cbc8548))
* **screen:** replace `utf8` icons with `ascii` characters ([de53255](https://github.com/mibe-iot/mirror/commit/de53255f34f4f0771d4e2081823b29c3b0cc5db5))

### [2.1.3](https://github.com/mibe-iot/mirror/compare/v2.1.2...v2.1.3) (2022-05-16)


### Bug Fixes

* **pairing:** replace `password` with `psk` ([db99c7f](https://github.com/mibe-iot/mirror/commit/db99c7f937a0a199fc9f9f695b66e097397f1d8f))

### [2.1.2](https://github.com/mibe-iot/mirror/compare/v2.1.1...v2.1.2) (2022-05-16)


### Bug Fixes

* **pairing:** shared context ([d60f334](https://github.com/mibe-iot/mirror/commit/d60f334adf654aeda246dbe6a6707950dd5cf0ca))

### [2.1.1](https://github.com/mibe-iot/mirror/compare/v2.1.0...v2.1.1) (2022-05-13)


### Bug Fixes

* **pairing:** bind this to characteristic initializer ([97e876d](https://github.com/mibe-iot/mirror/commit/97e876df9e1d41e093ee85d721cd05252c8a42a3))

## [2.1.0](https://github.com/mibe-iot/mirror/compare/v2.0.0...v2.1.0) (2022-05-12)


### Features

* **screen:** add subscribers & blessed log box ([a9f7ad8](https://github.com/mibe-iot/mirror/commit/a9f7ad8e65db027bb0c7f3fd9414c2f9931daa02))
* **screen:** create library for screen management ([cdd2e0f](https://github.com/mibe-iot/mirror/commit/cdd2e0f47da1d6aa21f431073e9375ba7f7fbe04))
* **screen:** create screen library ([#3](https://github.com/mibe-iot/mirror/issues/3)) ([82cf088](https://github.com/mibe-iot/mirror/commit/82cf0884bd6679ce6d9312e10ea0f7ad64e9c53e))
* **screen:** integrate library for screen management ([42c566c](https://github.com/mibe-iot/mirror/commit/42c566c293475fdc5be21bd573d76967748ffcbe))


### Bug Fixes

* **main:** remove mqtt subscribers from connected callback ([d5fd85d](https://github.com/mibe-iot/mirror/commit/d5fd85d28df7812969fd06ef3cb068bf8a095460))

## [2.0.0](https://github.com/mibe-iot/mirror/compare/v1.3.1...v2.0.0) (2022-05-11)


### ⚠ BREAKING CHANGES

* **deps:** added `node-wifi` package for working with Wi-Fi

### Features

* **main:** integrate business logic ([faec158](https://github.com/mibe-iot/mirror/commit/faec15823a511a190821ed282875db0fd8182cf2))
* **pairing:** add MQTT and Wi-Fi support ([a40e9f9](https://github.com/mibe-iot/mirror/commit/a40e9f9a56ac3034b6d85db5c9055d6dcdbb8a22))
* **pairing:** add Wi-Fi and MQTT connectors ([04d2239](https://github.com/mibe-iot/mirror/commit/04d22395473c77e8c9f977aecb9afe7785fc164e))


### Bug Fixes

* **config:** add Wi-Fi connected option for MQTT prepared ([8062d7f](https://github.com/mibe-iot/mirror/commit/8062d7f8a3fe21af0c229ad0684ad12dfee0eaa5))
* **workflows:** add branches to release workflow ([8f09258](https://github.com/mibe-iot/mirror/commit/8f09258e9b93cf046b8c04c34def6e80928c173d))


### Build System

* **deps:** add `node-wifi` module ([0d953e6](https://github.com/mibe-iot/mirror/commit/0d953e6115a7fa730640f8ac345bbe6b285a3b6e))

## [2.0.0-beta.3](https://github.com/mibe-iot/mirror/compare/v2.0.0-beta.2...v2.0.0-beta.3) (2022-05-11)


### Features

* **main:** integrate business logic ([faec158](https://github.com/mibe-iot/mirror/commit/faec15823a511a190821ed282875db0fd8182cf2))
* **pairing:** add Wi-Fi and MQTT connectors ([04d2239](https://github.com/mibe-iot/mirror/commit/04d22395473c77e8c9f977aecb9afe7785fc164e))

## [2.0.0-beta.2](https://github.com/mibe-iot/mirror/compare/v2.0.0-beta.1...v2.0.0-beta.2) (2022-05-11)


### Bug Fixes

* **config:** add Wi-Fi connected option for MQTT prepared ([8062d7f](https://github.com/mibe-iot/mirror/commit/8062d7f8a3fe21af0c229ad0684ad12dfee0eaa5))

## [2.0.0-beta.1](https://github.com/mibe-iot/mirror/compare/v1.3.1...v2.0.0-beta.1) (2022-05-11)


### ⚠ BREAKING CHANGES

* **deps:** added `node-wifi` package for working with Wi-Fi

### Bug Fixes

* **workflows:** add branches to release workflow ([8f09258](https://github.com/mibe-iot/mirror/commit/8f09258e9b93cf046b8c04c34def6e80928c173d))


### Build System

* **deps:** add `node-wifi` module ([0d953e6](https://github.com/mibe-iot/mirror/commit/0d953e6115a7fa730640f8ac345bbe6b285a3b6e))

### [1.3.1](https://github.com/mibe-iot/mirror/compare/v1.3.0...v1.3.1) (2022-05-11)


### Bug Fixes

* **workflows:** add branches to release workflow ([a9ac8c9](https://github.com/mibe-iot/mirror/commit/a9ac8c9f62e28ff6be9f90146bbc1c29ec6565cf))

## [1.3.0](https://github.com/mibe-iot/mirror/compare/v1.2.0...v1.3.0) (2022-05-11)


### Features

* **config:** add BLE app settings ([f289f08](https://github.com/mibe-iot/mirror/commit/f289f08d8bcd35a32882a0618e43713ac8d875de))
* **main:** integrate pairing module ([1f6900e](https://github.com/mibe-iot/mirror/commit/1f6900e713064a25b1c96bfa4967ea2d0d6b8b24))
* **pairing:** add pairing module ([7c95701](https://github.com/mibe-iot/mirror/commit/7c9570150e382845c6be11f842364239d475fe38))


### Bug Fixes

* **ble:** remove separated module ([a806710](https://github.com/mibe-iot/mirror/commit/a80671050d2ace25483ed3b0e714b845294635cd))
* **config:** change settings priority ([00d6e4c](https://github.com/mibe-iot/mirror/commit/00d6e4cfa8453e05429f29937b3c4a334d1ff47b))

## [1.2.0](https://github.com/mibe-iot/mirror/compare/v1.1.0...v1.2.0) (2022-05-11)



src/libs/config/config.js
/**
 * Application settings
 * @typedef {BLESettings & WifiSettings & MQTTSettings}
 */
export const Settings = {
  ...BLESettings,
  ...WifiSettings,
  ...MQTTSettings,
}

/**
 * WiFi Application Settings
 * @type {{WifiPrepared: boolean, WifiConnected: boolean, onWifiConnected: WifiSettings.onWifiConnected, SSID: string, Password: string, onWifiPrepared: WifiSettings.onWifiPrepared}}
 */
export const WifiSettings = {
  SSID: '',
  Password: '',
  WifiPrepared: false,
  WifiConnected: false,
  onWifiPrepared: () => {},
  onWifiConnected: () => {},
}

/**
 * MQTT Application Settings
 * @type {{Identifier: string, onMQTTPrepared: MQTTSettings.onMQTTPrepared, MQTTConnected: boolean, onMQTTConnected: MQTTSettings.onMQTTConnected, MQTTPrepared: boolean}}
 */
export const MQTTSettings = {
  Identifier: '',
  MQTTPrepared: false,
  MQTTConnected: false,
  onMQTTPrepared: () => {},
  onMQTTConnected: () => {},
}

/**
 * BLE Application Settings
 * @type {{onBLEConnected: onBLEConnected}}
 */
export const BLESettings = {
  /**
   * @callback onBLEConnected
   */
  onBLEConnected: () => {},
}

const preparedStateHandler = {
  set(target, prop, value, receiver) {
    const result = Reflect.set(...arguments)
    if (target.SSID && target.Password && !target.WifiPrepared) {
      target.WifiPrepared = true
      target.onWifiPrepared()
    }
    if (target.Identifier && !target.MQTTPrepared) {
      target.MQTTPrepared = true
      target.onMQTTPrepared()
    }
    return result
  },
}

export default new Proxy(Settings, preparedStateHandler)


src/libs/screen/index.js
import blessed from 'blessed'
import { defaultStyle } from './styles.js'
import { MqttClient } from 'mqtt'

const pairingSteps = ['BLE', 'Wi-Fi', 'MQTT']

export class Screen {
  constructor() {
    this.screen = blessed.screen({
      smartCSR: true,
    })
    this.screen.title = 'Mibe IoT Mirror'
    this.screen.key(['escape', 'q', 'C-c'], () => {
      return process.exit(0)
    })
  }

  /**
   * Set step configured
   * @param step {string}
   */
  pairingStepConfigured(step) {
    let flag = !pairingSteps.includes(step)
    this.pairingTable.setData(
      pairingSteps.map((value) => {
        if (!flag) {
          if (value === step) flag = true
          return [value, '+']
        }
        return [value, '-']
      })
    )
    this.screen.render()
  }

  initializePairingWindow() {
    this.pairingTable = blessed.table({
      height: '13%',
      width: '15%',
      top: 'center',
      left: 'center',
      ...defaultStyle,
    })
    this.screen.append(this.pairingTable)
    this.pairingTable.setLabel('Connections')
    this.pairingStepConfigured('')
  }

  /**
   * MQTT Window initializer
   * @param client {MqttClient}
   */
  initializeMqttWindow(client) {
    this.pairingTable.top = '0'
    this.pairingTable.left = '0'
    this.screen.render()

    const rpiApiUrl = blessed.text({
      top: '2%',
      left: '16%',
      content:
        'Raspberry Pi Server Api Endpoint: ' + process.env.RPI_SERVER_API_URL,
    })

    this.screen.append(rpiApiUrl)

    const infoBox = blessed.log({
      width: '100%',
      height: '87%',
      top: '13%',
      ...defaultStyle,
    })

    client.subscribe('#', (err) => {
      if (err) console.error(err)
    })

    client.on('message', (topic, message) => {
      infoBox.add(`Topic: ${topic}; Message: ${message.toString()}`)
    })

    this.screen.append(infoBox)
    this.screen.render()
  }
}
